### PR TITLE
Adding sections on addresses

### PR DESCRIPTION
**Pull request by gusthoff**
Adding sections: 
- addresses
- access and address

### CODE REVIEW

File: content/courses/advanced-ada/parts/data_types/types_representation.rst [hunk 1]
```diff
@@ -2754,6 +2761,402 @@ these array declarations are equivalent:
...
+pointer arithmetic.) The concept of addresses does exist in Ada, but it's
+mainly reserved for very specific applications, mostly related to low-level
+programming. In general, other approaches |mdash| such as using access types
+|mdash| are more than sufficient. (We discuss
+:doc:`access types <../resource_management/access_types>` in another chapter.
+Also, later on in that chapter, we discuss the
+:ref:`relation between access types and addresses <Adv_Ada_Access_Address>`.)
+In this section, we discuss some details about using addresses in Ada.
+
+We make use of the :ada:`Address` type, which is defined in :ada:`System`
```
Comment (pat-rogers) [inline content/courses/advanced-ada/parts/data_types/types_representation.rst:2780]: Probably "defined in" should be "defined in the"

File: content/courses/advanced-ada/parts/data_types/types_representation.rst [hunk 2]
```diff
@@ -2754,6 +2761,402 @@ these array declarations are equivalent:
...
+package, to handle addresses. In contrast to other programming languages (such
+as C or C++), an address in Ada isn't an integer value |mdash| its definition
+actually depends on the compiler implementation, but for now, let's consider
```
Comment (pat-rogers) [inline content/courses/advanced-ada/parts/data_types/types_representation.rst:2783]: It is true that "its definition actually depends on the compiler implementation" but that is driven directly by the hardware.

File: content/courses/advanced-ada/parts/data_types/types_representation.rst [hunk 3]
```diff
@@ -2754,6 +2761,402 @@ these array declarations are equivalent:
...
+it to be a private type.
```
Comment (pat-rogers) [inline content/courses/advanced-ada/parts/data_types/types_representation.rst:2784]: Do you want to justify it being a private type (usually, but not necessarily)? That justification being an attempt at application code portability, given the variations in hardware that result in different definitions of what an address actually is...

File: content/courses/advanced-ada/parts/data_types/types_representation.rst [hunk 4]
```diff
@@ -2754,6 +2761,402 @@ these array declarations are equivalent:
...
+.. code:: ada run_button project=Courses.Advanced_Ada.Data_Types.Type_Representation.Addresses.Address_Aspect
+
+    with System; use System;
+    with System.Address_Image;
+
+    with Ada.Text_IO; use Ada.Text_IO;
+
+    procedure Show_Address is
+
+       I_Main   : Integer;
```
Comment (pat-rogers) [inline content/courses/advanced-ada/parts/data_types/types_representation.rst:2871]: Should mark I_Main as aliased... but then the remaining examples below might look like they are incorrect.  Not sure what to really suggest. Maybe a note in prose?

File: content/courses/advanced-ada/parts/data_types/types_representation.rst [hunk 5]
```diff
@@ -2754,6 +2761,402 @@ these array declarations are equivalent:
...
+
+    with Ada.Text_IO; use Ada.Text_IO;
+
+    procedure Show_Address is
+
+       Arr : array (1 .. 10) of Integer;
+       A   : Address := Arr (1)'Address;
+       --               ^^^^^^^^^^^^^^^
+       --   Initializing address object with
+       --   address of the first component of Arr.
```
Comment (pat-rogers) [inline content/courses/advanced-ada/parts/data_types/types_representation.rst:3078]: Note that the language guarantees that Arr'Address = Arr(1)'Address

File: content/courses/advanced-ada/parts/resource_management/access_types.rst [hunk 6]
```diff
@@ -5963,3 +5970,138 @@ be legal, as we said before.
...
+       else
+          Put_Line ("AI_1 /= AI_2");
+       end if;
+    end Show_Address;
+
+Here, we convert the :ada:`A` address back to an access value by calling
+:ada:`To_Pointer (A)`. (When running this object, we see that :ada:`AI_1`
+and :ada:`AI_2` have the same access value.)
+
+.. admonition:: In the Ada Reference Manual
```
Comment (pat-rogers) [inline content/courses/advanced-ada/parts/resource_management/access_types.rst:6105]: Does the admonition mention that the conversions might not work, eg when the designated type is unbounded, such as String?

File: content/courses/advanced-ada/parts/data_types/types_representation.rst (hunk 2 above)
Comment (gusthoff) [reply content/courses/advanced-ada/parts/data_types/types_representation.rst:2783]: Thanks for the hint! I've just added this information to the paragraph.

File: content/courses/advanced-ada/parts/data_types/types_representation.rst (hunk 3 above)
Comment (gusthoff) [reply content/courses/advanced-ada/parts/data_types/types_representation.rst:2784]: Yes, that's a very good point. I've just extended the text and added most of what you've just said.

File: content/courses/advanced-ada/parts/data_types/types_representation.rst (hunk 4 above)
Comment (gusthoff) [reply content/courses/advanced-ada/parts/data_types/types_representation.rst:2871]: I've just marked `I_Main` as aliased. (I hope this doesn't become confusing for the readers.)

File: content/courses/advanced-ada/parts/data_types/types_representation.rst (hunk 5 above)
Comment (gusthoff) [reply content/courses/advanced-ada/parts/data_types/types_representation.rst:3078]: Thanks for the hint! I modified the code and added a comment about this equivalence.

File: content/courses/advanced-ada/parts/resource_management/access_types.rst (hunk 6 above)
Comment (gusthoff) [reply content/courses/advanced-ada/parts/resource_management/access_types.rst:6105]: Very good point! I created a subsection to discuss this — even though I wasn't able to write a code example that demonstrates problematic behavior.

**Review by pat-rogers — APPROVED** [review]

**Pull request merged**